core/src/main/java/cucumber/runtime/Runtime.java

```diff
@@ -75,29 +79,62 @@ public void run() {
         final StepDefinitionReporter stepDefinitionReporter = plugins.stepDefinitionReporter();
         runnerSupplier.get().reportStepDefinitions(stepDefinitionReporter);
 
+        final List<Future<?>> futures = new ArrayList<>();
         for (CucumberFeature feature : features) {
             for (final PickleEvent pickleEvent : feature.getPickles()) {
                 if (filters.matchesFilters(pickleEvent)) {
-                    executor.execute(new Runnable() {
+                    futures.add(executor.submit(new Runnable() {
                         @Override
                         public void run() {
                             runnerSupplier.get().runPickle(pickleEvent);
                         }
-                    });
+                    }));
                 }
             }
         }
+
         executor.shutdown();
-        try {
-            //noinspection StatementWithEmptyBody we wait, nothing else
-            while (!executor.awaitTermination(1, TimeUnit.DAYS)) ;
-        } catch (InterruptedException e) {
-            throw new CucumberException(e);
+
+        List<Throwable> thrown = new ArrayList<>();
+        for (Future f : futures) {
+            try {
+                f.get();
+            } catch (ExecutionException | InterruptedException e) {
+                thrown.add(e);
+            }
+        }
+        if(thrown.size() == 1){
+            throw new CucumberException(thrown.get(0));
+        } else if (thrown.size() > 1){
+            throw new MultipleFailureException(thrown);
         }
 
         bus.send(new TestRunFinished(bus.getTime(), bus.getTimeMillis()));
     }
 
+    class MultipleFailureException extends CucumberException {
+        private final List<Throwable> throwables;
+
+        MultipleFailureException(List<Throwable> throwables) {
+            super("There ");
+            this.throwables = throwables;
+        }
+
+        public List<Throwable> getFailures() {
+            return Collections.unmodifiableList(this.throwables);
+        }
+
+        public String getMessage() {
+            StringBuilder sb = new StringBuilder(String.format("There were %d exceptions:", this.throwables.size()));
+
+            for (Throwable e : this.throwables) {
+                sb.append(String.format("\n  %s(%s)", e.getClass().getName(), e.getMessage()));
```

Comment:
This doesn't log the stack traces for the original exceptions, does it?
Could make searching for the root cause difficult.

Reply:
Feel free to push improvements!